### ISSUE TITLE
t/basic.t fails with ~root = /

### ISSUE THREAD

**Issue by plicease** [dagolden/Path-Tiny#275]
```
not ok 73 - Test root homedir with longer path

#   Failed test 'Test root homedir with longer path'
#   at t/basic.t line 176.
#          got: '/foo/bar'
#     expected: '//foo/bar'
not ok 74 - ... and its stringification

#   Failed test '... and its stringification'
#   at t/basic.t line 177.
#          got: '/foo/bar'
#     expected: '//foo/bar'
not ok 75 - Test root homedir, longer path and trailing "/"

#   Failed test 'Test root homedir, longer path and trailing "/"'
#   at t/basic.t line 176.
#          got: '/foo/bar'
#     expected: '//foo/bar'
not ok 76 - ... and its stringification

#   Failed test '... and its stringification'
#   at t/basic.t line 177.
#          got: '/foo/bar'
#     expected: '//foo/bar'
```

Full install log here:

https://gist.github.com/plicease/3efb5158b62b34cf5651c2296406a9a5

And the verbose version of the test here:

https://gist.github.com/plicease/eae8cca9fddb24c9133cc567d7fd8bb1

perl -V:

https://gist.github.com/plicease/a2cacc9966cfce55e1479b50a328a00c

But I believe the most relevant detail is that this system has ~root = / and Path::Tiny appears to be doing the right thing, but the test doesn't account for this configuration.

**Comment by XSven** [comment]
We have detected the same failing tests using perl-5.14.4 on an AIX 7.2 operating system. If more details are needed I am willing to provide them.

**Comment by xdg** [comment]
@plicease and @XSven can you please try the https://github.com/dagolden/Path-Tiny/tree/root-expands-to-slash branch?

**Comment by plicease** [comment]
The root-expands-to-slash branch does indeed pass on the NetBSD system in which I encountered this.
https://gist.github.com/plicease/bcd30ae6904bc04993abe71e5a01fb4e

**Comment by ap** [comment]
@xdg I don’t like this fix too much. The fixup happens far away from the source of the value which at that point can have come from multiple different places. I would prefer to apply the fixup at the source, something like this:

````diff
diff --git i/t/basic.t w/t/basic.t
index 6ff1029005f..91700e17b3e 100644
--- i/t/basic.t
+++ w/t/basic.t
@@ -138,16 +138,20 @@ is $file->parent,  '/foo/baz';
     my $root_homedir = path(glob('~root'))->[0];
     my $missing_homedir = path(glob('~idontthinkso'))->[0];

+    # remove one trailing slash from a path string, if present
+    # so the result of concatenating a path that starts with a slash will be correct
+    sub S ($) { $_[0] =~ s!/\z!!; $_[0] }
+
     my @tests = (
       # [arg for path(), expected string (undef if eq arg for path()), test string]
         ['~',                     $homedir,                 'Test my homedir' ],
         ['~/',                    $homedir,                 'Test my homedir with trailing "/"' ],
-        ['~/foo/bar',             $homedir.'/foo/bar',      'Test my homedir with longer path' ],
-        ['~/foo/bar/',            $homedir.'/foo/bar',      'Test my homedir, longer path and trailing "/"' ],
+        ['~/foo/bar',           S($homedir).'/foo/bar',     'Test my homedir with longer path' ],
+        ['~/foo/bar/',          S($homedir).'/foo/bar',     'Test my homedir, longer path and trailing "/"' ],
         ['~root',                 $root_homedir,            'Test root homedir' ],
         ['~root/',                $root_homedir,            'Test root homedir with trailing /' ],
-        ['~root/foo/bar',         $root_homedir.'/foo/bar', 'Test root homedir with longer path' ],
-        ['~root/foo/bar/',        $root_homedir.'/foo/bar', 'Test root homedir, longer path and trailing "/"'],
+        ['~root/foo/bar',       S($root_homedir).'/foo/bar','Test root homedir with longer path' ],
+        ['~root/foo/bar/',      S($root_homedir).'/foo/bar','Test root homedir, longer path and trailing "/"'],
         ['~idontthinkso',         undef,                    'Test homedir of nonexistant user' ],
         ['~idontthinkso',         $missing_homedir,         'Test homedir of nonexistant user (via glob)' ],
         ['~blah blah',            undef,                    'Test space' ],
````

(I don’t know if the sub is named and/or written too tersely for your taste and whether you’d want to reindent the entire table, etc., hence why I didn’t submit this as a PR. Let me know if you want me to, along with any diverging preferences.)

**Comment by xdg** [comment]
@ap that approach (and name) would be fine with me. Can you submit it as a PR, please?

**Comment by shlomif** [comment]
hi all,

On Fri, 25 Nov 2022 07:24:21 -0800
Aristotle Pagaltzis ***@***.***> wrote:

> @xdg I don’t like this fix too much. The fixup happens far away from the
> source of the value which at that point can have come from multiple different
> places. I would prefer to apply the fixup at the source, something like this:
> 
> ````diff
> diff --git i/t/basic.t w/t/basic.t
> index 6ff1029005f..91700e17b3e 100644
> --- i/t/basic.t
> +++ w/t/basic.t
> @@ -138,16 +138,20 @@ is $file->parent,  '/foo/baz';
>      my $root_homedir = path(glob('~root'))->[0];
>      my $missing_homedir = path(glob('~idontthinkso'))->[0];
> 
> +    # remove one trailing slash from a path string, if present
> +    # so the result of concatenating a path that starts with a slash will be
> correct
> +    sub S ($) { $_[0] =~ s!/\z!!; $_[0] }
> +

AFAIK, this will mutate $_[0]'s input argument permanently. I suggest doing "my
($p) = @_;" instead.

>      my @tests = (
>        # [arg for path(), expected string (undef if eq arg for path()), test
> string] ['~',                     $homedir,                 'Test my homedir'
> ], ['~/',                    $homedir,                 'Test my homedir with
> trailing "/"' ],
> -        ['~/foo/bar',             $homedir.'/foo/bar',      'Test my homedir
> with longer path' ],
> -        ['~/foo/bar/',            $homedir.'/foo/bar',      'Test my
> homedir, longer path and trailing "/"' ],
> +        ['~/foo/bar',           S($homedir).'/foo/bar',     'Test my homedir
> with longer path' ],
> +        ['~/foo/bar/',          S($homedir).'/foo/bar',     'Test my
> homedir, longer path and trailing "/"' ], ['~root',
> $root_homedir,            'Test root homedir' ], ['~root/',
> $root_homedir,            'Test root homedir with trailing /' ],
> -        ['~root/foo/bar',         $root_homedir.'/foo/bar', 'Test root
> homedir with longer path' ],
> -        ['~root/foo/bar/',        $root_homedir.'/foo/bar', 'Test root
> homedir, longer path and trailing "/"'],
> +        ['~root/foo/bar',       S($root_homedir).'/foo/bar','Test root
> homedir with longer path' ],
> +        ['~root/foo/bar/',      S($root_homedir).'/foo/bar','Test root
> homedir, longer path and trailing "/"'], ['~idontthinkso',         undef,
>                'Test homedir of nonexistant user' ], ['~idontthinkso',
>  $missing_homedir,         'Test homedir of nonexistant user (via glob)' ],
> ['~blah blah',            undef,                    'Test space' ], ````
> 
> (I don’t know if the sub is named and/or written too tersely for your taste
> and whether you’d want to reindent the entire table, etc., hence why I didn’t
> submit this as a PR. Let me know if you want me to, along with any diverging
> preferences.)
> 



-- 

Shlomi Fish       https://www.shlomifish.org/
Escape from GNU Autohell - https://www.shlomifish.org/open-source/anti/autohell/

Where they have burned books, they will end in burning human beings.
    — https://en.wikiquote.org/wiki/Heinrich_Heine

Please reply to list if it's a mailing list post - https://shlom.in/reply .

**Comment by xdg** [comment]
Fixed by #276

**Comment by ap** [comment]
@shlomif:

> AFAIK, this will mutate $_[0]'s input argument permanently.

Thanks for catching that, I amended the PR to account for it.
